cpp/src/io/statistics/statistics_type_identification.cuh

```diff
@@ -0,0 +1,244 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/fixed_point/fixed_point.hpp>
+
+#include <cudf/wrappers/timestamps.hpp>
+
+#include <cudf/strings/string_view.cuh>
+
+#include <cudf/wrappers/durations.hpp>
+
+#include <cudf/utilities/traits.hpp>
+
+#include "conversion_type_select.cuh"
+
+#include <tuple>
+
+namespace cudf {
+namespace io {
+namespace detail {
+
+enum class io_type { ORC, PARQUET };
+
+template <io_type IO>
+struct conversion_map;
+
+// Every timestamp or duration type is converted to milliseconds in ORC statistics
+template <>
+struct conversion_map<io_type::ORC> {
```

Comment:
Apologies for wording this vaguely. Not a rhetorical question. I wanted to explore whether we could omit file type specific code from this PR. It's going to save a lot of loc and complexity.

Reply:
There is at least one obstacle towards that goal. ORC does not calculate unsigned integer or duration statistics. It is basically a no-op call. We could probably allow all the calculations to take place as it is and not do anything with the unneeded results in ORC. The class `statistics_type_category` was added because of that and can be removed if we do not care about doing extra work.

The other motivation of doing this conversion at the time of statistics calculation (and thus making the timestamp statistics processing dependent on format type) is this [issue](https://github.com/rapidsai/cudf/issues/6674). We could use `ts_scale` instead here and kick out `conversion_type_select.cuh` and `type_conversion` class.

Reply:
@vuule thoughts?

Reply:
The linked issue seems to refer to a different processing. IIUC, stats on timestamps are only min and max. The question is whether we can find the min and max during stats calculation and store it as a cudf type in `statistics_val`. Then convert just before writing it out. That conversion may use chrono.

Reply:
As for no-oping certain calculations, I think that maybe certain unneeded calculations are already happening https://github.com/rapidsai/cudf/pull/8191#discussion_r629255370. I may have misunderstood the template metaprogramming here

Reply:
> The linked issue seems to refer to a different processing. IIUC, stats on timestamps are only min and max. The question is whether we can find the min and max during stats calculation and store it as a cudf type in `statistics_val`. Then convert just before writing it out. That conversion may use chrono.

I think that right now it will not be possible because the kernel for writing out the statistics (at least in ORC) are not typed. We can remove this type based conversion from statistics and make that change later on if we clean up that code.

Reply:
> As for no-oping certain calculations, I think that maybe certain unneeded calculations are already happening [#8191 (comment)](https://github.com/rapidsai/cudf/pull/8191#discussion_r629255370). I may have misunderstood the template metaprogramming here

Yeah that was an oversight. It can be fixed in either direction at this point. I could remove the dependency on format completely and let redundant calculation be or I can edit the boolean condition to avoid aggregates for PARQUET.

Reply:
Oh so removal of stats_dtype from stats_enc.cu is planned in a future PR? Then it's ok.